tooling/minimize/src/vtable.rs

```diff
@@ -0,0 +1,108 @@
+use crate::*;
+
+impl<'tcx> Ctxt<'tcx> {
+    /// Gets the vtable name for the given type and trait object or creates it if it doesn't exist yet.
+    /// `trait_obj_ty` must be of kind [`rs::TyKind::Dynamic`].
+    pub fn get_vtable(&mut self, ty: rs::Ty<'tcx>, trait_obj_ty: rs::Ty<'tcx>) -> VTableName {
+        let rs::TyKind::Dynamic(trait_, _, _) = *trait_obj_ty.kind() else {
+            panic!("get_vtable called on non trait object type");
+        };
+        if let Some(vtable_name) = self.vtable_map.get(&(ty, trait_)) {
+            assert!(self.vtables.contains_key(*vtable_name));
+            *vtable_name
+        } else {
+            let fresh_name = VTableName(Name::from_internal(self.vtable_map.len() as _));
+            let vtable = self.generate_vtable(ty, trait_obj_ty);
+            self.vtables.insert(fresh_name, vtable);
+            self.vtable_map.insert((ty, trait_), fresh_name);
+            fresh_name
+        }
+    }
+
+    /// Generates a vtable for the given type and trait object.
+    fn generate_vtable(&mut self, ty: rs::Ty<'tcx>, trait_obj_ty: rs::Ty<'tcx>) -> VTable {
+        let rs::TyKind::Dynamic(trait_, _, _) = *trait_obj_ty.kind() else {
+            panic!("generate_vtable called on non trait object type");
+        };
+        // Get the size and align
+        let layout = self.rs_layout_of(ty);
+        assert!(layout.is_sized(), "There are no unsized trait objects");
+        let size = translate_size(layout.size);
+        let align = translate_align(layout.align.abi);
+
+        // Get the methods of the principal trait, create a method name wrapping the index in its vtable.
+        let methods = if let Some(trait_) = trait_.principal() {
+            let trait_ref = trait_.with_self_ty(self.tcx, ty);
+            let trait_ref = self.tcx.erase_regions(trait_ref);
+            let entries = self.tcx.vtable_entries(trait_ref);
+
+            entries
+                .iter()
+                .enumerate()
+                .filter_map(|(idx, entry)| self.vtable_entry_to_trait_method(idx as _, entry))
+                .collect()
+        } else {
+            Map::new()
+        };
+        let trait_name = self.get_trait_name(trait_obj_ty);
+
+        VTable { trait_name, size, align, methods }
+    }
+
+    /// Generate an method implementation for a vtable entry if it is a method.
+    fn vtable_entry_to_trait_method(
```

Comment:
This function is called only once -- I think the code might become a bit clearer even if you just inline it at its only call site.